tools/reactive-controllers/src/RovingTabindex.ts

```diff
@@ -0,0 +1,304 @@
+        }
+        const path = event.composedPath() as T[];
+        let targetIndex: number | undefined;
+        path.find((el) => {
+            targetIndex = this.elements.indexOf(el);
+            return targetIndex !== -1;
+        });
+        this.currentIndex = targetIndex ?? this.currentIndex;
+    };
+
+    handleFocusout = (event: FocusEvent): void => {
+        if (this.isRelatedTargetAnElement(event)) {
+            this.hostNoLongerContainsFocus();
+        }
+    };
+
+    acceptsEventCode(code: string): boolean {
+        if (code === 'End' || code === 'Home') {
+            return true;
+        }
+        switch (this.direction) {
+            case 'horizontal':
+                return code === 'ArrowLeft' || code === 'ArrowRight';
+            case 'vertical':
+                return code === 'ArrowUp' || code === 'ArrowDown';
+            case 'both':
+            case 'grid':
+                return code.startsWith('Arrow');
+        }
+    }
+
+    handleKeydown = (event: KeyboardEvent): void => {
+        if (!this.acceptsEventCode(event.code) || event.defaultPrevented) {
+            return;
+        }
+        let diff = 0;
+        switch (event.code) {
+            case 'ArrowRight':
+                diff += 1;
+                event.preventDefault();
+                break;
+            case 'ArrowDown':
+                diff += this.direction === 'grid' ? this.directionLength : 1;
+                event.preventDefault();
+                break;
+            case 'ArrowLeft':
+                diff -= 1;
+                event.preventDefault();
+                break;
+            case 'ArrowUp':
+                diff -= this.direction === 'grid' ? this.directionLength : 1;
+                event.preventDefault();
+                break;
+            case 'End':
+                this.currentIndex = 0;
+                diff -= 1;
+                event.preventDefault();
+                break;
+            case 'Home':
+                this.currentIndex = this.elements.length - 1;
+                diff += 1;
+                event.preventDefault();
+                break;
+        }
+        this.setCurrentIndexCircularly(diff);
+        // To allow the `focusInIndex` to be calculated with the "after" state of the keyboard interaction
+        // do `elementEnterAction` _before_ focusing the next element.
+        this.elementEnterAction(this.elements[this.currentIndex]);
+        this.focus();
+    };
+
+    manageTabindexes(): void {
+        if (this.focused) {
+            this.updateTabindexes(() => -1);
+        } else {
+            this.updateTabindexes((el: HTMLElement): number | false => {
+                const containsFocusInElement = el.contains(this.focusInElement);
+                if (el === this.focusInElement) {
+                    return 0;
+                } else if (containsFocusInElement) {
+                    return false;
+                }
+                return -1;
+            });
```

Comment:
This is a weird function signature (0, -1, or `false`). I don't think this is necessarily better, but wanted to think of an alternative:

```suggestion
            this.updateTabindexes((el: HTMLElement): number | undefined => {
                if (el === this.focusInElement) return 0;
                return el.contains(this.focusInElement) ? undefined : -1;
            });
```

Not really an improvement to `updateTabindexes` either; maybe we can simplify these later instead:

```ts
    updateTabindexes(getTabIndex: (el: HTMLElement) => number | undefined): void {
        this.elements.forEach((el) => {
            const tabIndex = getTabIndex(el);
            if ([0, -1].includes(tabIndex)) {
                el.tabIndex = tabIndex;
                return;
            }
            el.removeAttribute('tabindex');
            const updatable = el as unknown as {
                requestUpdate?: () => void;
            };
            if (updatable.requestUpdate) updatable.requestUpdate();
        });
    }
```